Bugfix: slot data stack is badly cleaned in templating

SamyCookie opened the pull request:
see ticket #9383 in trac. - https://twistedmatrix.com/trac/ticket/9383

In the templating module, more precisely in the '_flattenElement' method, when getting through elements in the tree, some cleanings of the slot data stack are missing in some cases.

Commit:
Bugfix: in templating module, slot data stack is badly cleaned after
flattening a node tree

adiroiban commented on src/twisted/web/newsfragments/9383.bugfix:
```diff
@@ -0,0 +1 @@
+in templating module, slot data stack is badly cleaned after flattening a node tree
```
Please start this with the full name of the public class/module/function which is affected...and end the sentence with a full stop :)

SamyCookie replied:
done

adiroiban commented on src/twisted/web/test/test_flatten.py:
```diff
@@ -431,6 +431,18 @@ def test_serializeDeferredSlots(self):
+    
+    
+    def test_slotDataPropagation(self):
+        """ Test that verify if a slot data inserted in a node is correctly propagated
```
See this for how to write docstrings https://jml.io/pages/test-docstrings.html

See the docstring from for the other tests.

No need for "test that", "check that". All tests are written to test something :)

What means "correctly"?

SamyCookie replied:
I tried to be more explicit !

adiroiban commented on src/twisted/web/test/test_flatten.py:
```diff
@@ -431,6 +431,18 @@ def test_serializeDeferredSlots(self):
+        span2 = tags.span( slot('test', default='17') )
+        tree = tags.div( tags.p(span1).fillSlots(test='42'), tags.p(span2) )
+        result = b'<div><p><span>42</span></p><p><span>23</span></p></div>'
+        return self.assertFlattensTo(tree, result)
```
See code formatting for the other Twisted code. http://twistedmatrix.com/documents/current/core/development/policy/index.html

I am not sure which is the system under test here. Which function or code is tested?

adiroiban commented on src/twisted/web/_flatten.py:
```diff
@@ -229,50 +229,51 @@ def keepGoing(newRoot, dataEscaper=dataEscaper,
         write(b'-->')
     elif isinstance(root, Tag):
         slotData.append(root.slotData)
+        
```
Minor comment
Do we need this change?
I know that you might not like the spacing..but other dev might not like your style...so I don't want to start an edit war :)

adiroiban commented on src/twisted/web/_flatten.py:
```diff
@@ -229,50 +229,51 @@ def keepGoing(newRoot, dataEscaper=dataEscaper,
-            return
-
-        if not root.tagName:
+        
```
I don't what to say about these changes.

I would prefer to have `return` here, and maybe use something like `yield keepGoingFinal(result)` which will call slotData.pop()

With the current changes, slotData.pop() is called after many new branches... so this would need more than 1 new test.

adiroiban reviewed (changes requested):
Thanks for looking into this issue.

See inline comments.

It is not clear what path / paths are tested by the new tests.

I see that slotData.pop() is not called after more than one execution paths... but I only see a new tests.

Please consider updating the docstring of the tests so that I know what is the intention of that test.

In the same time, the Appveyour and Travis tests are failing at twisted.web.test.test_flatten.SerializationTests.test_slotDataPropagation

which is the new tests that you wrote.

New and existing tests should pass across all the supported environments before merging or being able to approve this change.

Thanks!

SamyCookie commented:
@adiroiban sorry for the loooooooooooooong response.

I have updated my commits, removed the polemic one, and made the test pass (I don't know how I failed that).
Is this all more understandable ?

adiroiban commented on src/twisted/web/newsfragments/9383.bugfix:
```diff
@@ -0,0 +1 @@
+twisted.web._flatten._flattenElement, slot data stack is cleaned after flattening a node tree.
```
this talks about a private API.

What public API is targeted by this change?

SamyCookie replied:
`twisted.web.template.flatten` and `twisted.web.template.flattenString`

adiroiban commented on src/twisted/web/test/test_flatten.py:
```diff
@@ -433,6 +433,28 @@ def test_serializeDeferredSlots(self):
+        result = (
+            b'<div>'
+            b'<p><span>42</span></p>'
+            b'<p><span>17</span></p>'  # '42' was written before #9383
```
no need to document the broken implementation.
just try to explain what is the expected result.
so I think that this comment can be removed.

adiroiban commented on src/twisted/web/test/test_flatten.py:
```diff
@@ -433,6 +433,28 @@ def test_serializeDeferredSlots(self):
+        slot data propagation is only made to child nodes.
+        Before #9383 fix, slot datas were not correctly clean during tree
+        traversal and some elements that are not directly child nodes can
+        reach forbidden slot data.
```
What is a forbidden slot data?

I have never worked with this API...so I am not sure about the terminology.

But to me, it looks like this test is about.

```
Slot data is applied only to the targeted tag.
```

No need to talk about previous bugs or invalid behavior.
Just document the correct behavior as good as possible.

adiroiban commented on src/twisted/web/test/test_flatten.py:
```diff
@@ -433,6 +433,28 @@ def test_serializeDeferredSlots(self):
+        span1 = tags.span(slot('test'))
+        span2 = tags.span(slot('test', default='17'))
+        tree = tags.div(
+            tags.p(span1).fillSlots(test='42'),
```
not sure why we need the P tag for this test.
Is it really needed?

adiroiban reviewed (changes requested):
Thanks for the update.
Don't worry about the delay.
Looks good.

I still see 2 `slotData.pop()`, yet there is only a new test.

Is the new test covering both changes?

To make it easy to follow the implementation, I would expect to have 2 tests, each test with is own docstring.

I have never  used the flatten API so my review might be wrong.

My suggestion is to reach other Twisted developers ove IRC or mailling list and see if someone else can take a look at this.
